appcelerator/titanium_mobile
# [TIMOB-25061] Android: Fixed bug where TextField/TextArea setSelection() wrongly triggers a "change" event.

**JIRA:** https://jira.appcelerator.org/browse/TIMOB-25061

**Test:**

1. Create an app using the below code and run it on Android.
2. Tap the "Move Cursor to Front" button.
3. Verify the cursor is on the left side of the field.
4. Verify that a "change" event toast message is **NOT** displayed.
5. Tap the "Move Cursor to Back" button.
6. Verify the cursor is at the end of the field.
7. Verify that a "change" event toast message is **NOT** displayed.
8. Tap the "Select First Character" button.
9. Verify that the 1st character in the TextField is highlighted.
10. Verify that a "change" event toast message is **NOT** displayed.
11. Tap the "Select Last Character" button.
12. Verify that the last character in the TextField is highlighted.
13. Verify that a "change" event toast message is **NOT** displayed.

---

```javascript
var window = Ti.UI.createWindow({ layout: "vertical" });
var textField = Ti.UI.createTextField(
{
	value: "Hello World!",
	width: "90%",
	top: "5%",
});
textField.addEventListener("change", function(e) {
	var message = "TextField 'change' event received.";
	Ti.API.info(message);
	if ((Ti.Platform.name === "android") || (Ti.Platform.name === "windows")) {
		var notification = Ti.UI.createNotification(
		{
			message: message,
			duration: Ti.UI.NOTIFICATION_DURATION_SHORT,
		}).show();
	}
});
window.add(textField);
var moveCursorToFrontButton = Ti.UI.createButton(
{
	title: "Move Cursor to Front",
});
moveCursorToFrontButton.addEventListener("click", function(e) {
	textField.setSelection(0, 0);
});
window.add(moveCursorToFrontButton);
var moveCursorToBackButton = Ti.UI.createButton(
{
	title: "Move Cursor to Back",
});
moveCursorToBackButton.addEventListener("click", function(e) {
	var length = textField.value.length;
	textField.setSelection(length, length);
});
window.add(moveCursorToBackButton);
var selectFirstCharacterButton = Ti.UI.createButton(
{
	title: "Select First Character",
});
selectFirstCharacterButton.addEventListener("click", function(e) {
	var lastIndex = (textField.value.length > 0) ? 1 : 0;
	textField.setSelection(0, lastIndex);
});
window.add(selectFirstCharacterButton);
var selectLastCharacterButton = Ti.UI.createButton(
{
	title: "Select Last Character",
});
selectLastCharacterButton.addEventListener("click", function(e) {
	var length = textField.value.length;
	var startIndex = (length > 0) ? (length - 1) : 0;
	textField.setSelection(startIndex, length);
});
window.add(selectLastCharacterButton);
var selectAllButton = Ti.UI.createButton(
{
	title: "Select All",
});
selectAllButton.addEventListener("click", function(e) {
	textField.setSelection(0, textField.value.length);
});
window.add(selectAllButton);
window.open();
```
